### ISSUE TITLE
[ttnn/softmax]: float32 softmax defaults to the approximate exp: outputs up to 7.5% off, rows sum to 0.96–0.97, while ttnn.exp defaults to the accurate kernel

### ISSUE THREAD

**Issue by rybruscoe** [tenstorrent/tt-metal#57633]
### Component / Area

ops (ttnn/normalization/softmax)

### Issue Type (optional)

Bad Outputs

### Observed

With default arguments, `ttnn.softmax` on a `float32` tensor returns values that are wrong by up to **7.5% per element** (median 1.5%), and the rows do not sum to 1. The all-ones row is the clearest case: the answer is exactly `1/n`, `torch.softmax` returns it exactly, and the device returns `0.9709/n` at every row length from 32 to 4096 (0.9718 at 16384).

Blackhole, `[1, 1, 32, n]` float32, TILE, `dim=-1`; reference is the float64 softmax of the same float32 input; `torch.softmax` in float32 on the same input is shown for scale.

| input | n | torch fp32 rel max | **default** rel max | default rel median | default row sum | `math_approx_mode=False` rel max | its row sum |
|---|---|---|---|---|---|---|---|
| ones | 32…4096 | 0 | **2.91e-2** | 2.91e-2 | **0.9709** | 4.88e-4 | 0.9995 |
| ones | 16384 | 0 | 2.82e-2 | 2.82e-2 | 0.9718 | 4.88e-4 | 0.9995 |
| N(0,1) | 32 | 3.9e-7 | **6.45e-2** | 1.69e-2 | 0.9646–0.9998 | 3.3e-3 | 0.998–1.000 |
| N(0,1) | 4096 | 4.8e-7 | **7.45e-2** | 1.52e-2 | 0.9639–0.9999 | 4.9e-3 | 0.998–1.000 |
| N(0,1) | 16384 | 6.5e-7 | 6.56e-2 | 1.53e-2 | 0.9740–1.0017 | 4.8e-3 | 1.000–1.002 |
| 4·N(0,1) (logit-like) | 1024 | 1.5e-6 | **7.01e-2** | 2.32e-2 | 0.9617–1.0000 | 2.1e-2 | 0.998–1.000 |
| 4·N(0,1) | 16384 | 5.3e-6 | 6.98e-2 | 1.99e-2 | 0.9873–1.0021 | 2.3e-2 | 1.000–1.002 |
| U(0,1) | 4096 | 2.5e-7 | 4.07e-2 | 1.46e-2 | 0.9993–0.9999 | 2.2e-3 | 0.999–1.000 |

`numeric_stable=True` and a `compute_kernel_config` with `HiFi4 + fp32_dest_acc_en=True` change **no bit** of the default result, so this is not accumulation. Only `math_approx_mode=False` changes it.

The per-element error is the approximate exp's own error. On the same board, over `[-8, 0]` (the range the numerically-stable path feeds, `x - max ≤ 0`):

| kernel | rel error max | rel error median |
|---|---|---|
| `ttnn.exp(x)` (default = accurate) | 6.6e-8 | — |
| `ttnn.exp(x, fast_and_approximate_mode=True)` | **3.1e-2** | 2.0e-2 |

`exp_fast(0) = 0.9785` and `exp_fast(1)/e = 0.971`: the same −2.9% the ones row shows.

## Cause

`softmax_init_compute_kernel_config` hard-codes `default_approx_mode = true` for every input dtype:

```cpp
// ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_device_operation.cpp:386 (main e4a9afc5)
return init_device_compute_kernel_config(arch, compute_kernel_config, default_fidelity, true, is_fp32, false);
//                                                                                     ^^^^ math_approx_mode
```

Every softmax program factory then emits `compute_defines["EXP_APPROX"] = math_approx_mode ? "1" : "0"`, and the compute kernels call `exp_tile_init<EXP_APPROX>()` / `exp_tile<EXP_APPROX>()`. So a float32 tensor, which the same line deliberately gives an fp32 Dest (`is_fp32` → `fp32_dest_acc_en`), is exponentiated with the bf16-class approximate kernel. The unary op does the opposite: `ttnn.exp` documents `fast_and_approximate_mode ... Defaults to False`. Nothing in the `ttnn.softmax` docstring says the default is approximate; `docs/Softmax.md` lists `math_approx_mode=True` only under "Fast Approximation Configuration".

Same shape as #55130 (bias_gelu silently approximate while gelu defaults accurate) and #55265 (matmul activation silently approximate). #52045 (bf16 row sums) noted the unconditional `math_approx_mode = true` in passing as "a second, smaller defect on the same line ... not what this issue is about", measured through the row sum only (0.994 at w=4096); the per-element figure is ~15× larger than the row-sum residual because the exp error partially cancels between numerator and denominator. It has not been filed on its own.

## Residual, for completeness

With `math_approx_mode=False` the ones row is still 4.88e-4 low at every n (exp(0)=1 and the sum are exact, so this is `recip_tile`'s legacy Newton seed, #53106), and logit-scale inputs still reach 2e-2. Those are separate; this issue is only about the default selecting the approximate exp for float32.

### Expected

A float32 softmax uses the accurate exp by default, the same default ttnn.exp documents (fast_and_approximate_mode=False). For the all-ones row every element is exactly 1/n and rows sum to 1 within float32 rounding; for N(0,1) inputs per-element error is at the level of the accurate kernel (~1e-3 or better), not 7.5%.

### 1. Steps (exact commands)

Run the snippet below on a Blackhole device with the listed tt-metal commit:

```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
n = 4096
a = ttnn.from_torch(torch.ones(1, 1, 32, n), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
o = ttnn.to_torch(ttnn.softmax(a, dim=-1))
print(o[0, 0, 0, 0].item(), 1 / n, o[0, 0, 0].sum().item())   # 0.000237029046 0.000244140625 0.9708710
cfg = ttnn.WormholeComputeKernelConfig(math_fidelity=ttnn.MathFidelity.HiFi4, math_approx_mode=False, fp32_dest_acc_en=True, packer_l1_acc=False)
print(ttnn.to_torch(ttnn.softmax(a, dim=-1, compute_kernel_config=cfg))[0, 0, 0].sum().item())  # 0.9995
```

### 2. Input data / link or description

Generated in the snippet above (torch.ones / torch.randn with manual_seed(0)); no external data. The tables above were produced by the same script at the listed shapes.

### 3. Frequency

Always; deterministic across runs.

### 1. Software Versions

OS version: Ubuntu 22.04
Software / firmware release version: tt-metal a1c3bb93e (2026-09-16) with an unrelated cumsum patch; the softmax code path is byte-identical on main e4a9afc5 (2026-09-23).

### 2. Hardware Details

Product: Blackhole
Card/System: P150, single device

### Is this a regression?

Unknown

### Logs & Diagnostics

I have a fix (default the approximation off when the input is float32, mirroring the fp32-Dest policy on the same line; bf16 keeps today's default so model perf is untouched) validated on Blackhole with a regression test. Measured cost of the accurate exp on float32, median of 10 after 3 warm-ups:

| shape | today (approx exp) | accurate exp |
|---|---|---|
| [1,1,1024,4096] | 0.161 ms | 0.287 ms |
| [1,32,1024,1024] | 0.892 ms | 1.064 ms |
| [1,8,4096,4096] | 3.195 ms | 3.762 ms |

Anyone who wants today's speed on float32 keeps it with an explicit `math_approx_mode=True`, the same opt-in `ttnn.exp` uses. Happy to open the PR.

### Priority

P2

### Impact

Affected workflows: any float32 softmax caller, including attention reference paths, silently gets up to 7.5% per-element error and rows that sum to ~0.97.
Release or date risk: none known.
